modules/031-linstor/images/linstor-pools-importer/main.go

```diff
@@ -0,0 +1,593 @@
+	var opts options
+	opts.delaySeconds = *flag.Int("delay", 10, "Delay in seconds between scanning attempts")
+	klog.Infof("delaySeconds: %d", opts.delaySeconds)
+
+	opts.nodeName = os.Getenv("NODE_NAME")
+	if opts.nodeName == "" {
+		return opts, fmt.Errorf("Required NODE_NAME env variable is not specified!")
+	}
+	klog.Infof("nodeName: %s", opts.nodeName)
+
+	opts.podName = os.Getenv("POD_NAME")
+	if opts.podName == "" {
+		return opts, fmt.Errorf("Required POD_NAME env variable is not specified!")
+	}
+
+	klog.Infof("podName: %s", opts.podName)
+
+	opts.namespace = os.Getenv("NAMESPACE")
+	if opts.namespace == "" {
+		return opts, fmt.Errorf("Required NAMESPACE env variable is not specified!")
+	}
+	klog.Infof("namespace: %s", opts.namespace)
+	flag.Parse()
+	return opts, nil
+}
+
+// Create Kubernetes client
+func createKubeClient() (kclient.Client, error) {
+	var kc kclient.Client
+	clientConfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		clientcmd.NewDefaultClientConfigLoadingRules(),
+		&clientcmd.ConfigOverrides{},
+	)
+	// Get a config to talk to the apiserver
+	config, err := clientConfig.ClientConfig()
+	if err != nil {
+		return kc, fmt.Errorf("Failed to get kubernetes config: %w", err)
+	}
+	kc, err = kclient.New(config, kclient.Options{})
+	if err != nil {
+		return kc, fmt.Errorf("Failed to create Kubernetes client: %w", err)
+	}
+	return kc, nil
+}
+
+// Get owner
+func getOwner(ctx context.Context, kc kclient.Client, namespace, podName string) (v1.ObjectReference, error) {
+	var owner v1.ObjectReference
+	var pod v1.Pod
+	err := kc.Get(ctx, types.NamespacedName{Name: podName, Namespace: namespace}, &pod)
+	if err != nil {
+		return owner, fmt.Errorf("look up owner(s) of pod %s/%s: %v", namespace, podName, err)
+	}
+	owner = v1.ObjectReference{
+		APIVersion: "v1",
+		Kind:       "Pod",
+		Name:       pod.GetName(),
+		Namespace:  pod.GetNamespace(),
+		UID:        pod.GetUID(),
+	}
+	klog.Infof("using %s/%s as owner for Kubernetes events", owner.Kind, owner.Name)
+	return owner, nil
+}
+
+func provisionStoragePools(ctx context.Context, lc *lclient.Client, kc kclient.Client, owner *v1.ObjectReference, nodeName string, delay time.Duration) error {
+	candiCh := make(chan Candidate)
+	errCh := make(chan error)
+
+	go func() {
+		seen := make(map[uint32]struct{})
+
+		for {
+			candidates, err := getCandidates(nodeName)
+			if err != nil {
+				// only fatal error, cannot cmd
+				errCh <- err
+				return
+			}
+			for _, cand := range candidates {
+				if _, yes := seen[cand.UniqueKey]; yes {
+					continue
+				}
+				seen[cand.UniqueKey] = struct{}{}
+				candiCh <- cand
+			}
+			time.Sleep(delay)
+		}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+
+		case err := <-errCh:
+			return fmt.Errorf("Cannot get storage pool candidates: %w", err)
+
+		case cand := <-candiCh:
+			if cand.SkipReason != "" {
+				klog.Infof("Skip %s as it %s", cand.Name, cand.SkipReason)
+				continue
+			}
```

Comment:
The problem is that we must send it, because if we'd throw an error at the beginning and do not send it to the channel it will be repeated each new scanning, eg 10 seconds.
We need to remember which candidate it was or disable error reporting at all.

Otherwise we can move the seen logic into each driver implementation. eg. to have separate `seen` structure for `getLVMcandidates()` and `getLVMThinCandidates()`. But from my original design these functions should simple return parsed output of `lvs` and `vgs` commands.